v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/migrations/metadata/CassandraSourceMetadata.java

```diff
@@ -0,0 +1,236 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.cloud.teleport.v2.spanner.migrations.metadata;
+
+import com.datastax.oss.driver.api.core.cql.ResultSet;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.ColumnPK;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.NameAndCols;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.Schema;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceColumnDefinition;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceColumnType;
+import com.google.cloud.teleport.v2.spanner.migrations.schema.SourceTable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The {@code CassandraSourceMetadata} class is responsible for generating metadata from a Cassandra
+ * schema using a {@link ResultSet}, converting it into a Spanner-compatible format, and managing
+ * this metadata to facilitate schema migration operations.
+ *
+ * <p>This class supports the following functionalities:
+ *
+ * <ul>
+ *   <li>Extracting table, column, and primary key information from Cassandra's schema.
+ *   <li>Converting Cassandra schema details into Spanner-compatible objects like {@link
+ *       SourceTable}.
+ *   <li>Updating a provided {@link Schema} instance with the extracted metadata.
+ * </ul>
+ *
+ * <p>The metadata extraction process uses the {@link ResultSet} containing the schema details from
+ * Cassandra, such as table names, column definitions, data types, and primary key details.
+ *
+ * <p><strong>Note:</strong> This class does not perform direct database interactions; it relies on
+ * a pre-populated {@link ResultSet}.
+ */
+public class CassandraSourceMetadata {
+
+  private final Schema schema;
+  private final ResultSet resultSet;
+
+  /**
+   * Private constructor to initialize {@link CassandraSourceMetadata}.
+   *
+   * @param resultSet The {@link ResultSet} containing Cassandra schema metadata. Cannot be null.
+   * @param schema The {@link Schema} instance to update with metadata. Cannot be null.
+   */
+  private CassandraSourceMetadata(ResultSet resultSet, Schema schema) {
+    this.resultSet = Objects.requireNonNull(resultSet, "ResultSet cannot be null");
+    this.schema = Objects.requireNonNull(schema, "Schema cannot be null");
+  }
+
+  /**
+   * Generates a map of table names to {@link SourceTable} objects, representing the schema of the
+   * Cassandra source in a Spanner-compatible format.
+   *
+   * @return A map where keys are table names and values are {@link SourceTable} objects containing
+   *     schema details.
+   */
+  public Map<String, SourceTable> generateSourceSchema() {
```

Comment:
this method can be marked as private / default.

Reply:
Resolved @bharadwaj-aditya